src/rpc.cpp

```diff
@@ -43,10 +41,54 @@ static const char * const RealmObjectTypesFunction = "function";
 static const char * const RealmObjectTypesList = "list";
 static const char * const RealmObjectTypesObject = "object";
 static const char * const RealmObjectTypesResults = "results";
+static const char * const RealmObjectTypesRealm = "realm";
 static const char * const RealmObjectTypesUndefined = "undefined";
 
+static RPCServer*& get_rpc_server(JSGlobalContextRef ctx) {
+    static std::map<JSGlobalContextRef, RPCServer*> s_map;
+    return s_map[ctx];
+}
+
+RPCWorker::RPCWorker() {
+    m_thread = std::thread([this]() {
+        // TODO: Create ALooper/CFRunLoop to support async calls.
+        while (!m_stop) {
+            try_run_task();
+        }
+    });
+}
+
+RPCWorker::~RPCWorker() {
+    m_stop = true;
+    m_thread.join();
+}
+
+void RPCWorker::add_task(std::function<json()> task) {
+    m_tasks.push_back(std::packaged_task<json()>(task));
+}
+
+json RPCWorker::pop_task_result() {
+    auto future = m_futures.pop_back();
+    return future.get();
+}
+
+void RPCWorker::try_run_task() {
+    try {
+        // Use a 10 millisecond timeout to keep this thread unblocked.
+        auto task = m_tasks.pop_back(10);
+        task();
+
+        // Since this can be called recursively, it must be pushed to the front of the queue *after* running the task.
+        m_futures.push_front(task.get_future());
+    }
+    catch (ConcurrentDequeTimeout &) {
+        // We tried.
```

Comment:
Without it the thread can deadlock, we need the condition to timeout in order to keep the loop going and checking its underlying condition. It's a similar concept to when you use `[NSRunLoop runUntilDate:]`.